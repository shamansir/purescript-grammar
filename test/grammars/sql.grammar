main :- repSep(statementSemicolon, ws).

statementSemicolon :- [statement, ";"].
statement :- (selectStmt | createTableStmt).

selectStmt :- [selectKW, ws, selection, ws, fromKW, ws, tableName].
selection :- repSep(columnName, commaWS).

createTableStmt :- [createKW, ws, tableKW, ws, tableName, ws, "(", ws, repSep(colSpec, commaWS), ws, ")"].

colSpec :- [columnName, ws, type, ws, (refClause | "")].
refClause :- [refKW, ws, tableName, ws, "(", columnName, ")"].

type :- ("INT" | "TEXT").

columnName :- (ident | placeholder).
tableName :- (ident | placeholder).

createKW :- "CREATE".
tableKW :- "TABLE".
selectKW :- "SELECT".
fromKW :- "FROM".
refKW :- "REFERENCES".

ident :- [alpha, repSep((alphaNum | "."), "")].
string :- ["\"", repSep(stringChar, ""), "\""].
stringChar :- (^'"' | ['\\', '"']).
alpha :- ([a-z] | [A-Z] | "_").
num :- [0-9].
alphaNum :- (alpha | num).
ws :- repSep((" "|"\n"), "").
commaWS :- [",", ws].
placeholder :- "???".
