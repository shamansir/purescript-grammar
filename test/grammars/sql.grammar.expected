main :- repSep(statementSemicolon,ws).
alpha :- ([a-z]|[A-Z]|"_").
alphaNum :- (alpha|num).
colSpec :- [columnName,ws,type,ws,(refClause|"")].
columnName :- (ident|placeholder).
commaWS :- [",",ws].
createKW :- "CREATE".
createTableStmt :- [createKW,ws,tableKW,ws,tableName,ws,"(",ws,repSep(colSpec,commaWS),ws,")"].
fromKW :- "FROM".
ident :- [alpha,repSep((alphaNum|"."),"")].
num :- [0-9].
placeholder :- "???".
refClause :- [refKW,ws,tableName,ws,"(",columnName,")"].
refKW :- "REFERENCES".
selectKW :- "SELECT".
selectStmt :- [selectKW,ws,selection,ws,fromKW,ws,tableName].
selection :- repSep(columnName,commaWS).
statement :- (selectStmt|createTableStmt).
statementSemicolon :- [statement,";"].
string :- ["\"",repSep(stringChar,""),"\""].
stringChar :- (^'"'|['\\','"']).
tableKW :- "TABLE".
tableName :- (ident|placeholder).
type :- ("INT"|"TEXT").
ws :- repSep((" "|"\n"),"").